Define runtime compile time types (#21658)
dotnet/docs: docs/csharp/programming-guide/types/index.md

## Implicit types, anonymous types, and nullable value types

As stated previously, you can implicitly type a local variable (but not class members) by using the [var](../../language-reference/keywords/var.md) keyword. The variable still receives a type at compile time, but the type is provided by the compiler. For more information, see [Implicitly Typed Local Variables](../classes-and-structs/implicitly-typed-local-variables.md).

It can be inconvenient to create a named type for simple sets of related values that you don't intend to store or pass outside method boundaries. You can create *anonymous types* for this purpose. For more information, see [Anonymous Types](../classes-and-structs/anonymous-types.md).

> Ordinary value types can't have a value of [null](../../language-reference/keywords/null.md). However, you can create nullable value types by affixing a `?` after the type. For example, `int?` is an `int` type that can also have the value [null](../../language-reference/keywords/null.md). Nullable value types are instances of the generic struct type <xref:System.Nullable%601?displayProperty=nameWithType>. Nullable value types are especially useful when you're passing data to and from databases in which numeric values might be null. For more information, see [Nullable value types](../../language-reference/builtin-types/nullable-value-types.md).

Reviewer: ```suggestion
Ordinary value types can't have a value of [null](../../language-reference/keywords/null.md). However, you can create nullable value types by appending a `?` after the type. For example, `int?` is an `int` type that can also have the value [null](../../language-reference/keywords/null.md). Nullable value types are instances of the generic struct type <xref:System.Nullable%601?displayProperty=nameWithType>. Nullable value types are especially useful when you're passing data to and from databases in which numeric values might be null. For more information, see [Nullable value types](../../language-reference/builtin-types/nullable-value-types.md).
```

## Compile-time type and runtime type

> Any variable may have distinct compile time and runtime types. The *compile-time type* is the declared or inferred type of the variable in the source code. The *runtime type* is the type of the instance referred to by that variable. Often those two types are the same, as in the following example:

Reviewer: ```suggestion
A variable can have different compile-time and run-time types. The *compile-time type* is the declared or inferred type of the variable in the source code. The *run-time type* is the type of the instance referred to by that variable. Often those two types are the same, as in the following example:
```

See https://styleguides.azurewebsites.net/Styleguide/Read?id=2719&topicid=31304.

:::code language="csharp" source="snippets/index/Program.cs" id="CompileTimeType":::

> In other cases, the declared type is different, as shown in the following two examples:

Reviewer: ```suggestion
In other cases, the compile-time type is different, as shown in the following two examples:
```

:::code language="csharp" source="snippets/index/Program.cs" id="RuntimeTypes":::

> In both of the preceding examples, the runtime type is a `string`. The compile-time type is `object` in the first line, and `IEnumerable<char>` in the second.

Reviewer: ```suggestion
In both of the preceding examples, the run-time type is a `string`. The compile-time type is `object` in the first line, and `IEnumerable<char>` in the second.
```

> It's important to understand when the compile-time type and the runtime type apply when the two types are different for a variable. The compile-time type determines all the actions taken by the compiler. These compiler actions include method call resolution, overload resolution, available implicit and explicit casts. The runtime type determines all actions that are resolved at runtime. These runtime actions include virtual method call dispatch, `is` expressions and `switch` expression evaluation, and other type testing APIs. Understand which type applies to which operation to better understand how your code interacts with types.

Reviewer: ```suggestion
If the two types are different for a variable, it's important to understand when the compile-time type and the run-time type apply. The compile-time type determines all the actions taken by the compiler. These compiler actions include method call resolution, overload resolution, and available implicit and explicit casts. The run-time type determines all actions that are resolved at run time. These run-time actions include dispatching virtual method calls, evaluating `is` and `switch` expressions, and other type testing APIs. To better understand how your code interacts with types, recognize which action applies to which type.
```
